README.md
# paco-ua-api

## API for Portal Académico Online - Universidade de Aveiro

[![Latest Release](https://img.shields.io/github/v/release/digas99/paco-ua-api?label=Latest%20Version)](https://img.shields.io/github/v/release/digas99/paco-ua-api?label=Latest%20Version)

This API uses [Puppeteer](https://pptr.dev/) to, through a headless browser, log in and scrape the page **Secretaria Virtual** from [paco.ua.pt](https://paco.ua.pt). The results are routed with [Express.js](https://expressjs.com/).

This app is a personal project and has no affiliation with Universidade de Aveiro.

![logo](/public/paco-api-logo-45.png)

### [API Documentation](docs/README.md)

### Table of Contents

1. [Setup](#setup)
1. [Usage](#usage)
1. [Features](#features)
1. [Apps](#apps)

## Setup

### Setup the server locally with Docker:

Make sure to have Docker running on your machine. [[How to here]](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-18-04)

- Build the image

```
docker build --tag pacoua-api .
```

- Run the container

```
docker run -p 8000:8000 --name pacoua-api pacoua-api
```

---

(To run on a different port, change the value of the port on the left side (***8000***:8000) in the command above. The port on the right ride (8000:***8000***) has to always match the port in the file static.js)

### Setup the server locally manually:

- Install the node modules

```
$ npm install
```

- Run the app

```
$ npm run devStart
```

(To run on a different port, change the value of **PORT** in the file static.js)

After any of the two configurations, the server will be running at http://127.0.0.1:8000

## Usage

Every endpoint is accessed through a GET HTTP Request, with your institutional email credentials in the **Basic Authorization** Header (encrypted in base64). [More Information](https://en.wikipedia.org/wiki/Basic_access_authentication#:~:text=password%20(see%20below).-,Client%20side,-%5Bedit%5D)

```shell
GET /schedule
  -H 'accept: application/json'
  -H 'Authorization: Basic {base64_encrypted_credentials}'
```

If everything is correct, this should be the response:

```json5
// RESPONSE
{
    "data": {
        "schedule": {
            "Segunda": [
                {
                    "subject": {
                        "name": "SEGURANÇA INFORMÁTICA E NAS ORGANIZAÇÕES",
                        "abbrev": "SIO",
                        "code": "42573"
                    },
                    "start": "9h",
                    "duration": "2h",
                    "capacity": 24,
                    "class": "P8",
                    "room": "04.2.03"
                },
                {
                    "subject": {
                        "name": "ARQUITETURA DE COMPUTADORES I",
                        "abbrev": "AC-I",
                        "code": "41948"
                    },
                    "start": "14h",
                    "duration": "1h",
                    "capacity": 98,
                    "class": "TP1",
                    "room": "ANF. V"
                },
                ...
            ],
            "Terça": [
                {
                    "subject": {
                        "name": "ARQUITETURA DE COMPUTADORES I",
                        "abbrev": "AC-I",
                        "code": "41948"
                    },
                    "start": "9h",
                    "duration": "2h",
                    "capacity": 18,
                    "class": "P07",
                    "room": "04.2.17"
                },
                ...
            ],
            ...
        },
        "school_year": "2021/2022",
        "semester": 1
    },
    "url": "https://paco.ua.pt/secvirtual/horarios/c_horario_aluno.asp",
    "title": "Horário",
    "timestamp": "2022-08-16T01:51:09.193Z"
}
```

The same applies for all the other endpoints.

## Features

- Dados Pessoais &nbsp;&nbsp;`GET /personal`
- Situação de prescrição &nbsp;&nbsp;`GET /expiration`
- Histórico Notas &nbsp;&nbsp;`GET /subjects/history`
- Disciplinas Inscritas &nbsp;&nbsp;`GET /subjects/current`
- Estados das Propinas &nbsp;&nbsp;`GET /tuition_fees`
- Plano Curricular &nbsp;&nbsp;`GET /subjects`
- Calendário de Exames do Aluno &nbsp;&nbsp;`GET /exams`
- Calendário de Exames por Disciplina &nbsp;&nbsp;`GET /exams?subjects=...`
- Apoio às Aulas &nbsp;&nbsp;`GET /classes`
- Horário &nbsp;&nbsp;`GET /schedule`
- Requerimentos &nbsp;&nbsp;```GET /requests```

## Apps

If you have an app that uses this API, feel free to contact me or make a pull request of this README with your app, obeying the table format bellow.

| Logo | Name | Contributors | Description |
| ----- | ----- | ----- | ----- |
| <img width="32px" src="https://github.com/digas99/schedule-ua/blob/main/images/logo_32x32.png"></img> | [SchedUA](https://github.com/digas99/schedule-ua) | [@digas99](https://github.com/digas99) | Browser Extension for easy access to your Schedule from Universidade de Aveiro. |

docs/CHANGELOG.md
# [Changelog v1.1.0](https://github.com/digas99/paco-ua-api/releases/tag/v1.1.0)
Released on 27/06/2023

## Content
- Bypass of "Inquerito Pedagógico"

## Bug Fixes
- Improved the creation of the abbreviation of a subject when fetching the schedule of a specific subject

# [Changelog v1.0.8](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.8)
Released on 14/01/2023

## Bug Fixes
- Fixed issue where schedule couldn't be fetched if there was a message with overlapping classes

# [Changelog v1.0.7](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.7)
Released on 25/09/2022

## Bug Fixes
- Fixed the way the abbreviation of a subject is created when fetching the schedule of a specific subject

# [Changelog v1.0.6](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.6)
Released on 25/09/2022

## Bug Fixes
- Changed EventEmitter max listeners to prevent heroku from crashing server when multiple concurrent requests are made to the API

# [Changelog v1.0.5](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.5)
Released on 25/09/2022

## Content
- Added subjects codes to /schedule endpoint

# [Changelog v1.0.4](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.4)
Released on 23/09/2022

## Content
- Possibility to filter endpoint /tuition_fees by year
- Added Exams and Classes endpoints to swagger documentation

# [Changelog v1.0.3](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.3)
Released on 07/09/2022

## Bug Fixes
- Updated Schedule Parser to fix bug in the /schedule endpoint

# [Changelog v1.0.2](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.2)
Released on 04/09/2022

## Content
- Added Server Timeout (504) HTTP response to avoid crashing the app
- Added Forbidden Access (403) when the credentials provided in the GET Requests header are wrong 

# [Changelog v1.0.1](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.1)
Released on 26/08/2022

## Bug Fixes
- Missing enabling CORS for preflight operations

# [Changelog v1.0.0](https://github.com/digas99/paco-ua-api/releases/tag/v1.0.0)
Released on 23/08/2022

## Content
- API endpoints are accessbile through https://pacoua-api.pt
- API covers all the main sections from PACO-UA:
  - Dados Pessoais &nbsp;&nbsp;`GET /personal`
  - Situação de prescrição &nbsp;&nbsp;`GET /expiration`
  - Histórico Notas &nbsp;&nbsp;`GET /subjects/history`
  - Disciplinas Inscritas &nbsp;&nbsp;`GET /subjects/current`
  - Estados das Propinas &nbsp;&nbsp;`GET /tuition_fees`
  - Plano Curricular &nbsp;&nbsp;`GET /subjects`
  - Calendário de Exames do Aluno &nbsp;&nbsp;`GET /exams`
  - Calendário de Exames por Disciplina &nbsp;&nbsp;`GET /exams?subjects=...`
  - Apoio às Aulas &nbsp;&nbsp;`GET /classes`
  - Horário &nbsp;&nbsp;`GET /schedule`
  - Requerimentos &nbsp;&nbsp;```GET /requests```
- Interactive API documentation with Swagger is accessible through https://pacoua-api.pt/docs

static.js
module.exports = {
    // port for development
    PORT: 3001,

    // /personal
    PERSONAL_URL: "https://paco.ua.pt/secvirtual/c_dadospess.asp",
    PERSONAL_TITLE: "Dados Pessoais",

    // /subjects
    CURRENT_URL: "https://paco.ua.pt/secvirtual/c_examesInscr.asp",
    CURRENT_TITLE: "Disciplinas Inscritas",
    HISTORY_URL: "https://paco.ua.pt/secvirtual/c_historiconotas.asp",
    HISTORY_TITLE: "Histórico Notas",
    CURRICULUM_URL: "https://paco.ua.pt/secvirtual/c_planocurr.asp",
    CURRICULUM_TITLE: "Plano Curricular",

    // /tuition-fees
    TUITION_FEES_URL: "https://paco.ua.pt/secvirtual/c_estadoDasProprinas.asp",
    TUITION_FEES_TITLE: "Estados das Propinas",

    // /schedule
    SCHEDULE_URL: "https://paco.ua.pt/secvirtual/horarios/c_horario_aluno.asp",
    SCHEDULE_TITLE: "Horário",

    // exams
    EXAMS_URL: "https://paco.ua.pt/secvirtual/c_calendarioDeExames.asp",
    EXAMS_TITLE: "Calendário de Exames do Aluno",
    EXAMS_SUBJECTS_URL: "https://paco.ua.pt/secvirtual/c_calendarioDeExamesPorDisciplina.asp",
    EXAMS_SUBJECTS_TITLE: "Calendário de Exames por Disciplina",

    // /requests
    REQUESTS_URL: "https://paco.ua.pt/secvirtual/pedidos_requerimentos_consultas_new.asp",
    REQUESTS_TITLE: "Requerimentos",

    // /expiration
    EXPIRATION_URL: "https://paco.ua.pt/secvirtual/c_situacaoprescricao.asp",
    EXPIRATION_TITLE: "Situação de prescrição",

    // directories
    ROUTES_DIR: "./routes/",

    // classes
    CLASSES_URL: "https://paco.ua.pt/secvirtual/aulas/lista_turmas_aluno.asp",
    CLASSES_TITLE: "Apoio às Aulas",

    // miscelaneous
    MONTHS: {
        "janeiro": "01",
        "fevereiro": "02",
        "março": "03",
        "abril": "04",
        "maio": "05",
        "junho": "06",
        "julho": "07",
        "agosto": "08",
        "setembro": "09",
        "outubro": "10",
        "novembro": "11",
        "dezembro": "12",
    },
}